src/crawler/crawler.ts

```diff
@@ -0,0 +1,261 @@
+	): void {
+		this.process = processFn;
+	}
+
+	public async start(): Promise< void > {
+		if ( this.state.isActive ) {
+			this.log( 'log', 'Crawler already running' );
+			return;
+		}
+
+		this.state.isActive = true;
+		this.log( 'log', 'Crawler started' );
+
+		while ( this.state.isActive ) {
+			const next = await this.getNextUrl();
+			if ( next ) {
+				await this.processUrl( next );
+			} else {
+				this.state.isActive = false;
+				this.log( 'log', 'Crawler finished' );
+			}
+		}
+	}
+
+	private async processUrl( url: string ): Promise< void > {
+		this.log( 'log', 'processing url', url );
+		try {
+			// Wait until we're allowed to process the next URL
+			await this.waitForRateLimit();
+
+			await this.navigateToUrl( url );
+
+			// @TODO: Get the HTML content via bus?
+			const html = document.documentElement.outerHTML;
+
+			// Process the page content
+			await this.process( html );
+
+			// Extract and queue new URLs
+			const links = this.extractLinks( html );
+			await this.queueUrls( links, url );
+		} catch ( error ) {
+			this.log( 'error', 'Error processing URL', url, error );
+			this.state.isActive = false;
+		}
+	}
+
+	private async waitForRateLimit(): Promise< void > {
+		const now = Date.now();
+		const delayMs = 1000 / this.state.rateLimit; // Convert rate limit to milliseconds between requests
+
+		if ( now < this.state.nextProcessTime ) {
+			await new Promise( ( resolve ) =>
+				setTimeout( resolve, this.state.nextProcessTime - now )
+			);
+		}
+
+		// Calculate next allowed process time using the delay
+		this.state.nextProcessTime = now + delayMs;
+	}
+
+	private extractLinks( htmlString: string ): string[] {
+		// Create a DOM parser instance
+		const parser = new DOMParser();
+
+		// Parse the HTML string into a document
+		const doc = parser.parseFromString( htmlString, 'text/html' );
+
+		// Find all anchor tags
+		const linkElements = doc.querySelectorAll( 'a' );
+
+		// Convert NodeList to Array and extract link data
+		const links = Array.from( linkElements ).map( ( link ) => {
+			// Get the href attribute
+			const href = link.getAttribute( 'href' );
+
+			// Skip if no href, or it's a javascript: link or anchor link
+			if (
+				! href ||
+				href.startsWith( 'javascript:' ) ||
+				href.startsWith( '#' )
+			) {
+				return null;
+			}
+
+			// Try to resolve relative URLs to absolute
+			let absoluteUrl;
+			try {
+				absoluteUrl = new URL( href, window.location.origin ).href;
+			} catch ( e ) {
+				// If URL parsing fails, use the original href
+				absoluteUrl = href;
+			}
+
+			const isExternal = link.hostname !== window.location.hostname;
+			if ( isExternal ) {
+				return null;
+			}
+
+			return absoluteUrl;
+		} );
+
+		// Filter out null values and return unique links
+		return links
+			.filter( ( link ) => link !== null )
+			.filter(
+				( link, index, self ) =>
+					index === self.findIndex( ( l ) => l === link )
+			);
+	}
+
+	private async queueUrls(
```

Comment:
We're building [just that](https://github.com/WordPress/wordpress-playground/blob/571128bde1c2caf586d0462904952434ddaad30a/packages/playground/data-liberation/src/import/WP_Attachment_Downloader.php#L114) in the PHP plugin! :-) There are concurrent requests, resource limits, and if we run it in Playground, we'll still do the downloads via fetch() which means we benefit from authorized cookies.